core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java

```diff
@@ -0,0 +1,402 @@
+import alluxio.grpc.BlockMasterWorkerServiceGrpc;
+import alluxio.grpc.ConfigProperty;
+import alluxio.grpc.GrpcChannel;
+import alluxio.grpc.GrpcChannelBuilder;
+import alluxio.grpc.GrpcServer;
+import alluxio.grpc.GrpcServerAddress;
+import alluxio.grpc.GrpcServerBuilder;
+import alluxio.grpc.GrpcService;
+import alluxio.grpc.RegisterWorkerPRequest;
+import alluxio.grpc.RegisterWorkerPResponse;
+import alluxio.grpc.Scope;
+import alluxio.grpc.ServiceType;
+import alluxio.security.authentication.AuthType;
+import alluxio.security.user.ServerUserState;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.Channel;
+import io.grpc.stub.StreamObserver;
+import org.junit.After;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+public class RegisterStreamerTest {
+  private static final InetSocketAddress TEST_ADDRESS = new InetSocketAddress("localhost", 9999);
+
+  @Rule
+  public ConfigurationRule mConfigurationRule =
+      new ConfigurationRule(
+          new ImmutableMap.Builder<PropertyKey, Object>()
+          // set response timeout short to ensure that tests don't take too long
+          .put(PropertyKey.WORKER_REGISTER_STREAM_RESPONSE_TIMEOUT, 500)
+          .put(PropertyKey.WORKER_REGISTER_STREAM_DEADLINE, 2000)
+          .put(PropertyKey.WORKER_REGISTER_STREAM_COMPLETE_TIMEOUT, 500)
+          // set batch size to 1 so that the stream will send a request
+          // for every block, used for easily control the concurrency of the stream
+          .put(PropertyKey.WORKER_REGISTER_STREAM_BATCH_SIZE, 1)
+          // we don't handle authentication in testing
+          .put(PropertyKey.SECURITY_AUTHENTICATION_TYPE, AuthType.NOSASL)
+          .build(),
+          Configuration.modifiableGlobal());
+
+  private GrpcServer mServer;
+  private GrpcChannel mChannel;
+
+  @Test(timeout = 5000)
+  public void testRegisterTimeOutConcurrentRequest() throws Exception {
+    // create a server that does not respond to requests
+    createMockService(
+        new BlockMasterWorkerServiceGrpc.BlockMasterWorkerServiceImplBase() {
+          @Override
+          public StreamObserver<RegisterWorkerPRequest> registerWorkerStream(
+              StreamObserver<RegisterWorkerPResponse> responseObserver) {
+            return new StreamObserver<RegisterWorkerPRequest>() {
+              @Override
+              public void onNext(RegisterWorkerPRequest value) {
+              }
+
+              @Override
+              public void onError(Throwable t) {
+              }
+
+              @Override
+              public void onCompleted() {
+              }
+            };
```

Comment:
You can create an base class `NoOpStreamObserver` with empty impls for each method, and then extends this class in each test, overriding the methods that actually need to do something.